tools/ordered_set.go

```diff
@@ -0,0 +1,217 @@
+package tools
+
+// OrderedSet is a unique set of strings that maintains insertion order.
+type OrderedSet struct {
+	// s is the set of strings that we're keeping track of.
+	s []string
+	// m is a mapping of string value "s" into the index "i" that that
+	// string is present in in the given "s".
+	m map[string]int
```

Comment:
If `s` is always an ordered set of strings, why bother tracking each string's index in `m`?